llm/src/mosaic_gpt.py

```diff
@@ -86,10 +90,80 @@ def __init__(self, cfg: DictConfig, device: Optional[str] = None):
         )
         self.mhsa.out_proj._is_residual = True
 
-    def forward(self, x, key_padding_mask):
-        return self.mhsa(x,
-                         key_padding_mask=key_padding_mask,
-                         need_weights=False)
+    def forward(self, x, key_padding_mask, attn_mask=None):
+        assert attn_mask is None
+        return self.mhsa(
+            x,
+            key_padding_mask=key_padding_mask,
+            need_weights=False)
+
+    @staticmethod
+    def mask_shape(*args, **kwargs):
+        return None
+
+    @staticmethod
+    def attn_mask_(*args, **kwargs):
+        return None
+
+
+class TritonFlashCausalAttention(nn.Module):
+    """Multi-headed self attention using triton FlashAttn kernel which includes bias for Alibi integration
+    """
+
+    def __init__(self, cfg: DictConfig, device: Optional[str] = None):
+        super().__init__()
+        try:
+            from src.flash_attention import FlashMHA  # type: ignore
+        except ImportError as e:
+            raise e
+        
+        self.mhsa = FlashMHA(
+            embed_dim=cfg.d_model,
+            num_heads=cfg.n_heads,
+            attention_dropout=cfg.attn_pdrop,
+            bias=True,
+            batch_first=True,
+            causal=True,
+            device=device,
+        )
+        self.mhsa.out_proj._is_residual = True
+
+    def forward(self, x, key_padding_mask=None, attn_mask=None):
+        assert key_padding_mask is None
+        return self.mhsa(
+            x,
+            key_padding_mask=None,
+            attn_mask=attn_mask,
+            need_weights=False)
+
+    @staticmethod
+    def mask_shape(n_heads, seq_len, alibi):
+        return (1, n_heads, 1, seq_len) if alibi else (1, 1, 1, seq_len)
+
+    @staticmethod
+    def attn_mask_(attn_mask, n_heads, seq_len, alibi=False, alibi_bias_max=8):
+        # in-place fill causal attn mask
+        attn_mask.zero_()
+
+        if alibi:
+            device, dtype = attn_mask.device, attn_mask.dtype
+            attn_mask.add_(alibi_bias(n_heads, seq_len, full=False, alibi_bias_max=alibi_bias_max, device=device, dtype=dtype))
+
+        return attn_mask
+
+
+def alibi_bias(n_heads, seq_len, full=False, alibi_bias_max=8, device=None, dtype=None):
+    alibi_bias = torch.arange(1 - seq_len, 1, dtype=dtype, device=device).view(1, 1, 1, seq_len)
+    if full:
+        # generate 1 x Heads x SeqLen x SeqLen alibi bias mask
+        # otherwise the mask is 1 x Heads x 1 x SeqLen (which is braodcasted up to the approproate size)
+        alibi_bias = alibi_bias - torch.arange(1 - seq_len, 1, dtype=dtype, device=device).view(1, 1, seq_len, 1)
+        alibi_bias.abs_().mul_(-1)  # since we're using causal flag, this isn't really needed, but why not include it
+
+    m = torch.arange(1, n_heads + 1, dtype=dtype, device=device)
+    m.mul_(alibi_bias_max / n_heads)
+    alibi_bias = alibi_bias * (1. / (2 ** m.view(1, n_heads, 1, 1)))
```

Comment:
I didn't spend enough time doing the mental computations to figure out how this obtains the same values for _m_ as in the Composer implementation/paper, but your results look good so I'm not too concerned.

Reply:
Last sentence of 2nd paragraph of section 3 of [the alibi paper](https://arxiv.org/abs/2108.12409) goes over the details. This is my interpretation and convergence works well.